src/datumaro/plugins/data_formats/arrow/mapper/media.py

```diff
@@ -149,58 +116,64 @@ def forward(
     ) -> Dict[str, Any]:
         out = super().forward(obj)
 
-        _bytes = None
-        if isinstance(encoder, Callable):
-            _bytes = encoder(obj)
-        else:
-            _bytes = cls.encode(obj, scheme=encoder)
-        out["bytes"] = _bytes
+        _bytes = encoder(obj) if isinstance(encoder, Callable) else cls.encode(obj, scheme=encoder)
 
-        out["attributes"] = DictMapper.forward(dict(size=obj.size))
+        path = None if _bytes is not None else getattr(obj, "path", None)
 
+        out["image"] = {
+            "has_bytes": _bytes is not None,
+            "bytes": _bytes,
+            "path": path,
+            "size": obj.size,
+        }
         return out
 
     @classmethod
-    def backward(cls, obj: Dict[str, Any]) -> Image:
-        media_dict = cls.backward_dict(obj)
-
-        path = media_dict["path"]
-        attributes, _ = DictMapper.backward(media_dict["attributes"], 0)
-
-        _bytes = None
-        if media_dict.get("bytes", None) is not None:
-            _bytes = media_dict["bytes"]
+    def backward(
+        cls,
+        media_struct: pa.StructScalar,
+        idx: int,
+        table: pa.Table,
+        table_path: str,
+    ) -> Image:
+        image_struct = media_struct.get("image")
+
+        if path := image_struct.get("path").as_py():
+            return Image.from_file(
+                path=path,
+                size=image_struct.get("size").as_py(),
+            )
 
-        if _bytes:
-            return Image.from_bytes(data=_bytes, size=attributes["size"])
-        return Image.from_file(path=path, size=attributes["size"])
+        return Image.from_bytes(
+            data=lambda: pa.ipc.open_file(pa.memory_map(table_path, "r"))
+            .read_all()
+            .column("media")[idx]
+            .get("image")
+            .get("bytes")
+            .as_py(),
+            size=image_struct.get("size").as_py(),
+        )
 
     @classmethod
-    def backward_from_batches(
-        cls,
-        batches: List[pa.lib.RecordBatch],
-        parent: Optional[str] = None,
-    ) -> List[Image]:
-        paths = pa_batches_decoder(batches, f"{parent}.path" if parent else "path")
-        attributes_ = pa_batches_decoder(
-            batches, f"{parent}.attributes" if parent else "attributes"
+    def backward_extra_image(
+        cls, media_struct: pa.StructScalar, idx: int, table: pa.Table, extra_image_idx: int
+    ) -> Image:
+        image_struct = media_struct
```

Comment:
Minor comment: do we need to have this?

Reply:
Removed it and renamed `media_struct` with `image_struct` in the function argument.